// tweet-truncator/lib/index.js
// LICENSE : MIT
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.truncate = exports.TweetTruncator = void 0;
const tweet_truncator_1 = __importDefault(require("./tweet-truncator"));
exports.TweetTruncator = tweet_truncator_1.default;
/**
 * truncate contents object with maxLength.
 * @param {{
 *   desc?: string,
 *   quote?: string,
 *   title?: string,
 *   url?: string,
 *   tags?: string[]
 * }} contents
 * @param {object} options
 * @returns {string}
 */
function truncate(contents, options = {}) {
    const maxLength = options.maxLength || 280; // default: 280
    const twitterTr = new tweet_truncator_1.default(options);
    const fixedContents = {
        desc: "",
        quote: "",
        title: "",
        url: "",
        tags: [],
        ...contents
    };
    fixedContents.tags = fixedContents.tags.map((tag) => `#${tag}`);
    const status = twitterTr.joinContents(fixedContents);
    let over = twitterTr.getTweetLength(status) - maxLength;
    if (over > 0) {
        return twitterTr.truncateStatus(fixedContents, over);
    }
    return status;
}
exports.truncate = truncate;
